pkg/api/image/formatter.go

```diff
@@ -53,6 +54,33 @@ type Handler struct {
 	BackingImageCache           ctllhv1.BackingImageCache
 }
 
+func (h Handler) IsMatchedResource(resource subresource.Resource, method string) bool {
+	if resource.Name != subresource.VirtualMachineImages.Resource {
+		return false
+	}
+
+	if resource.SubResource == actionDownload && method == http.MethodGet {
+		return true
+	}
+
+	if resource.SubResource == actionUpload && method == http.MethodPost {
```

Comment:
Shouldn't this be HTTP method `PUT`, since the action is idempotent?

Reply:
What I'm doing is to migrate whole same functionality to sub resource handler, and original method is `POST`.  I didn't make too much changes to original behavior just in case, including HTTP method.

https://github.com/harvester/harvester/blob/bf8bdcd7423508f767bf389e43ee8e812639e01a/pkg/api/image/formatter.go#L71-L75

And the reason why it's `POST` because we're using rancher/steve to build our API server. In original documentation, `action` query string seems to only allow POST action ([reference](https://github.com/rancher/steve?tab=readme-ov-file#action)).

However, I was thinking and testing about this during these few days. I also think `PUT` would be better for those cases. So, I will use `PUT` method to replace all `POST` method if behavior is idempotent. I think this change would make our API look like more k8s standardized API. Otherwise, it's quite confused, which all APIs are `POST` method.

But, I'll still check all behavior is idempotent or not.

Reply:
Actually, because subresource concept is the verb to original resource, it's basically a `PUT` or `GET` method. There shouldn't be `POST` method here. It's that we use subresources (verbs) to manipulate existing resource instead of creating.